src/Storages/ObjectStorage/DataLakes/Iceberg/IcebergMetadata.cpp

```diff
@@ -376,6 +428,38 @@ void IcebergMetadata::updateSnapshot(ContextPtr local_context, Poco::JSON::Objec
             configuration_ptr->getPathForRead().path);
 }
 
+bool IcebergMetadata::optimize(const StorageMetadataPtr & metadata_snapshot, ContextPtr context, const std::optional<FormatSettings> & format_settings)
+{
+    if (context->getSettingsRef()[Setting::allow_experimental_iceberg_compaction])
+    {
+        auto configuration_ptr = configuration.lock();
+        const auto sample_block = std::make_shared<const Block>(metadata_snapshot->getSampleBlock());
+        compactIcebergTable(object_storage, configuration_ptr, format_settings, sample_block, context, true);
+        return true;
+    }
+    else
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Enable 'allow_experimental_iceberg_compaction' setting to call optimize for iceberg tables.");
+    }
+}
+
+void IcebergMetadata::scheduleBackgroundCompaction(ObjectStoragePtr object_storage_, ContextPtr context, const std::optional<FormatSettings> & format_settings_, SharedHeader sample_block)
+{
+    if (context->getSettingsRef()[Setting::allow_experimental_iceberg_background_compaction].value)
+    {
+        auto configuration_ptr = configuration.lock();
+
+        if (configuration_ptr->isDataLakeConfiguration() && configuration_ptr->supportsWrites())
```

Comment:
This is redundant here now, as it is already `IcebergMetadata` for which both of these conditions are true